### DISCUSSION TITLE
Strongswan IKEv1 connection to Cisco VPN Concentrator fails

### DISCUSSION THREAD

**Discussion by tomglx** [strongswan/strongswan#1637] [Q&A]
Hi,
we're trying to migrate a working ipsec-tools (racoon) configuration to strongswan v5.9.10 on Fedora 37.
Some VPNs are working some not. The non-working ones are solely ones having Cisco devices as targets.
After a security proposal is successfully chosen, the NAT negotiation determines that the remote gateway
is behind nat.  So far I understand it. It proceeds with an new negotiation over port 4500 for NAT and goes
into an timeout. racoon isn't doing this. Mainly because the remote GW is reachable without NAT.
What could be the reason for this and how to we prevent strongswan from believing that NAT is in use?
Here's our configuration. IPs and secret are obfuscated. A log from a connection trial is attached.

connections {
    itzbund-cvk1 {
        version = 1
        local_addrs = x.x.x.x
        remote_addrs = 1.1.1.1
        proposals = aes256-sha1-modp1024, default
        rekey_time = 24h
        reauth_time = 0
        local {
            auth = psk
        }
        remote {
            auth = psk
            id = 1.1.1.1
        }

        children {
            itzbund2cvk1 {
                esp_proposals = aes-sha1, default
                local_ts = 0.0.0.0/0,::/0
                remote_ts = 0.0.0.0/0,::/0
                rekey_time = 1h
                mode = tunnel
                start_action = start
                updown = "/etc/strongswan/updown.sh"
            }
        }
    }
}
secrets {
    ike-itzbund-cvk1 {
        secret = "xxxx"
        id = 1.1.1.1
    }
}
[itzbund.log](https://github.com/strongswan/strongswan/files/11138103/itzbund.log)

**Comment by tobiasbrunner** [comment]
> What could be the reason for this and how to we prevent strongswan from believing that NAT is in use?

There could be a static NAT on the remote end (port-forwarding or some 1:1 IP mapping). All strongSwan does is compare the contents of the server's NAT-D notify in the response with the one calculated from that packet's source IP/port. If that doesn't match there is a NAT on the responder's side (unless it created the value incorrectly e.g. by selecting the wrong source IP to calculate the hash).  From the log you posted, I can tell that instead of the public IP `62.x.x.x` it used the private IP `10.135.55.233` when generating the hash. So either there actually is a NAT to that IP or the server incorrectly used that IP for some reason.

Anyway, if the responder supports NAT-T (i.e. it responds with NAT-D notifies) it should be listening on port 4500 as per [RFC 3947, section 4](https://datatracker.ietf.org/doc/html/rfc3947#section-4). If there is a firewall between the two peers, it should allow such traffic. If there is any port-forwarding involved, make sure to also forward port 4500 besides port 500.

Note that it's not possible to disable NAT-T/D in strongSwan, but maybe Cisco has such an option (or perhaps it could use the public IP when calculating the hash even if that IP is not local to avoid that a NAT is detected by strongSwan).

**Comment by tomglx** [comment]
One last question remains. Port 4500 is in fact reachable from our gateway and there is no known firewall between the peers. 
Also in our old racoon configuration NAT-T negotiation for this peer is enabled and IKE remains on port 500 without NAT-T/D.
While for other peers, we explicitly have to disable it there.
Why does the NAT-T negotiation fail? Is it possible, that this can be a routing issue to the internal address 10.135.55.233 that is
announced by the peer?

**Reply by tobiasbrunner** [reply]
> Port 4500 is in fact reachable from our gateway and there is no known firewall between the peers.

How did you test this? Your gateway is the host on which strongSwan is running?

> Also in our old racoon configuration NAT-T negotiation for this peer is enabled and IKE remains on port 500 without NAT-T/D.

So NAT-D notifies are exchanged but no NAT is detected? (Do you have a detailed log of that?) Or what do you mean?

> Is it possible, that this can be a routing issue to the internal address 10.135.55.233 that is
> announced by the peer?

The peer does not announce that IP, that is, strongSwan does not know anything about it. I just brute forced the hash the peer sent to determine the IP it used to calculate it (again it either used that IP by mistake or there actually is a NAT involved). You have to check what happens at the remote end when packets arrive on port 4500 at the public IP strongSwan uses.

**Comment by tomglx** [comment]
Yes. My gateway is the host on which Strongswan is running. It has direct internet contact. So no NAT-T support is needed on our side.
I've just disabled NAT-T in racoon for testing. This makes no difference. The VPN runs fine. Even with NAT-T enabled, racoon doesn't even
make an attempt to use port 4500. tcpdump proves this. I can only assume, that racoon may ignore NAT-D messages at all?
For comparison I've attached the debug2 log of our racoon. The real remote addresses of interest are 62.153.205.229 and 195.243.136.117.
They're for an active and a backup gateway.
I have no administrative control over the remote end. It is operated by the german customs agency (ZIVIT). They've even certified 
strongswan as a viable VPN software for this. I can only assume that they've done this with a much older version and the old ipsec.conf syntax. They've checked my swanctl configuration and found no errors. They've only suggested to change the remote GW id to the internal address of their gateway. But this makes also no difference.
As for testing the NAT-T connectivity. This works.

nc -v -s 194.39.189.12 -u 62.153.205.229 4500
Ncat: Version 7.93 ( https://nmap.org/ncat )
Ncat: Connected to 62.153.205.229:4500.
[racoon.log](https://github.com/strongswan/strongswan/files/11156779/racoon.log)

**Reply by tobiasbrunner** [reply]
I don't see any NAT-T vendor IDs or NAT-D payloads getting exchanged in that log (e.g. there is no `nptype=20(nat-d)` or `nptype=130(nat-d)` anywhere). So it's no surprise that there is no switch to port 4500.

What's interesting is that there seem to be two connections initiated by those remote gateways, but none initiated by racoon. And these requests don't contain NAT-T VIDs. So if NAT-T is disabled on the remote end (at least as initiator), it's weird that it would respond with a NAT-T VID when strongSwan sends one. Do you have a log where racoon is the initiator and actually sends NAT-T VIDs? And what happens if strongSwan is the responder?

> I have no administrative control over the remote end. It is operated by the german customs agency (ZIVIT).

I see. Do they really only provide IKEv1?

> They've even certified strongswan as a viable VPN software for this. 

Do you have any documentation on that?

> I can only assume that they've done this with a much older version and the old ipsec.conf syntax.

The latter makes no difference. The version might if they used a very very old version (i.e. < 5.0).

> They've checked my swanctl configuration and found no errors. They've only suggested to change the remote GW id to the internal address of their gateway. But this makes also no difference.

No, that makes no difference in regards to NAT-T. Did the provide any information in regards to port 4500 on their end? Should that work? If not (e.g. their firewall config prevents it), they should probably consider disabling NAT-T in their implementation as it's not actually supported.

> As for testing the NAT-T connectivity. This works.

Hm, not sure how accurate that test is. `nc` will simply send a bunch of single-byte UDP packets to that port. If a firewall does not actually send back an ICMP (Destination/Port unreachable) but simply drops the packets (or such ICMPs get dropped by a firewall on the way), this will still be reported as success.

**Comment by tomglx** [comment]
We solved the problem. Root Cause was in fact the PSK. It contains a single quote and a backslash. Although we've specified it in double quotes, it was somehow misinterpreted internally by strongswan. Once we Hex-encoded it, the connection worked instantly.

**Reply by tobiasbrunner** [reply]
When using double quotes, backslashes have to be escaped because they are used to escape other characters (e.g. newlines in multi-line strings, and they can also be used to explicitly encode e.g. newlines or tabs via `\n` or `\t`). Without double quotes, it probably would have worked as neither backslashes nor single quotes are treated specially.
